src/core/Akka.Streams/Implementation/ChannelSinks.cs

```diff
@@ -0,0 +1,170 @@
+            _onWriteReady = OnWriteReady;
+            SetHandler(_inlet, this);
+        }
+
+        private void OnWriteFailed(Exception cause) => TryFail(cause);
+
+        private void OnWriteAvailable(bool available)
+        {
+            if (available && _writer.TryWrite(_awaitingElement))
+                Pull(_inlet);
+            else
+                TryComplete();
+        }
+
+        private void TryComplete()
+        {
+            CompleteStage();
+            if (_isOwner)
+                _writer.TryComplete();
+        }
+
+        private void TryFail(Exception e)
+        {
+            FailStage(e);
+            if (_isOwner)
+                _writer.TryComplete(e);
+        }
+
+        public override void OnUpstreamFinish()
+        {
+            base.OnUpstreamFinish();
+            if (_isOwner)
+                _writer.TryComplete();
+        }
+
+        public override void OnUpstreamFailure(Exception e)
+        {
+            base.OnUpstreamFailure(e);
+            if (_isOwner)
+                _writer.TryComplete(e);
+        }
+
+        public override void PreStart()
+        {
+            Pull(_inlet);
+            base.PreStart();
+        }
+
+        public override void OnPush()
+        {
+            var element = Grab(_inlet);
+            if (_writer.TryWrite(element))
+            {
+                Pull(_inlet);
+            }
+            else
+            {
+                var continuation = _writer.WaitToWriteAsync();
+                if (continuation.IsCompletedSuccessfully)
+                {
+                    var available = continuation.GetAwaiter().GetResult();
+                    if (available && _writer.TryWrite(element))
+                    {
+                        Pull(_inlet);
+                    }
+                    else
+                    {
+                        TryComplete();
+                    }
+                }
+                else
+                {
+                    var task = continuation.AsTask();
+                    _awaitingElement = element;
+                    task.ContinueWith(_onWriteReady);
+                }
+            }
+
+        }
+
+        private void OnWriteReady(Task<bool> t)
+        {
+            if (t.IsFaulted) _onWriteFailed(t.Exception);
+            else if (t.IsCanceled)
+                _onWriteFailed(new TaskCanceledException(t));
+            else _onWriteAvailable(t.Result);
+        }
+    }
+    
+    internal sealed class ChannelReaderSink<T> : GraphStageWithMaterializedValue<SinkShape<T>, ChannelReader<T>>
+    {
+        private readonly int _bufferSize;
+        private readonly bool _singleReader;
+        private readonly BoundedChannelFullMode _fullMode;
+
+        public ChannelReaderSink(int bufferSize, bool singleReader = false, BoundedChannelFullMode fullMode = BoundedChannelFullMode.Wait)
+        {
+            _bufferSize = bufferSize;
+            _singleReader = singleReader;
```

Comment:
Shouldn't `singleReader` be `true` by default?

Reply:
This was 'as-found' but `singleReader = false` here is the safer default so I didn't think it was worth changing.

Reply:
FWIW, `BoundedChannel` AFAIK doesn't even use `ChannelOptions.SingleReader`. it's only used in `UnboundedChannel`

Reply:
And we don't use `UnboundedChannel`, so I guess this setting doesn't really matter then?

Reply:
> And we don't use `UnboundedChannel`, so I guess this setting doesn't really matter then?

Not _yet_ anyway. The challenge of course is that it may become relevant in the future.

Right now, `BoundedChannel<T>` uses a `Dequeue<T>` internally, so it uses a single lock for both readers/writers. In the future however, the internal structure may change (e.x. something more like `ConcurrentQueue<T>` with segments) and then it may become important.

Really it's a question of whether it is an oversight in `ChannelOptions` that it is exposed for both types.